app/src/main/java/com/humberto/tasky/auth/presentation/login/LoginViewModel.kt

```diff
@@ -0,0 +1,63 @@
+package com.humberto.tasky.auth.presentation.login
+
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.compose.runtime.snapshotFlow
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.humberto.tasky.auth.domain.AuthRepository
+import com.humberto.tasky.auth.domain.UserDataValidator
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class LoginViewModel @Inject constructor(
+    private val authRepository: AuthRepository,
+    private val userDataValidator: UserDataValidator
+): ViewModel() {
+
+    var state by mutableStateOf(LoginState())
+        private set
+
+    private val eventChannel = Channel<LoginEvent>()
+    val events = eventChannel.receiveAsFlow()
+
+    init {
+        viewModelScope.launch {
+            collectCombinedEmailAndPassword()
+        }
+    }
+
+    private suspend fun collectCombinedEmailAndPassword() {
+        combine(
+            snapshotFlow { state.email.text },
+            snapshotFlow { state.password.text }
+        ) { email, password ->
+            val isValidEmail = userDataValidator.isValidEmail(email.toString())
+            state = state.copy(
+                isEmailValid = isValidEmail,
+                canLogin = isValidEmail &&
+                        !state.isLoggingIn &&
+                        password.isNotEmpty()
+            )
+        }.collectLatest { }
```

Comment:
Could it make more sense to use `launchIn(viewModelScope)` so that this flow’s lifecycle is automatically tied to the ViewModel? This way it can be cancelled when the ViewModel is cleared